### PR TITLE
feat: migrate lightgbmranker to mono

### PR DESCRIPTION
**Pull request by bwanglzu**
1. move lightgbm ranker to executors repo.
2. unify lightgbm with catboost, make them similar to use

### CODE REVIEW

File: jinahub/rankers/LightGBMRanker/lightgbm_ranker.py [hunk 1]
```diff
@@ -1,8 +1,204 @@
...
+          and `match_features` will be extracted from corresponded matches documents tags w.r.t the query document.
+
+        :param docs: :class:`DocumentArray` passed by the user or previous executor.
+        :param kwargs: Additional key value arguments.
+        """
+        if os.path.exists(self.model_path):
+            dataset = self._get_features_dataset(docs)
+            predictions = self.booster.predict(dataset.get_data())
+            for prediction, match in zip(predictions, docs.traverse_flat(['m'])):
+                match.scores[self.label] = prediction
```
Comment (JoanFM) [inline jinahub/rankers/LightGBMRanker/lightgbm_ranker.py:196]: remember to `sort` matches as in `catboost`

**Review by JoanFM — CHANGES_REQUESTED** [review]

File: jinahub/rankers/LightGBMRanker/LICENSE [hunk 2]
```diff
@@ -0,0 +1,193 @@
+Copyright 2020-2021 Jina AI Limited.  All rights reserved.
```
Comment (tadejsv) [inline jinahub/rankers/LightGBMRanker/LICENSE:1]: Do we need this? We have a license in this repository already

File: jinahub/rankers/LightGBMRanker/README.md [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+
+- [🌱 Prerequisites](#-prerequisites)
+- [🚀 Usages](#-usages)
+- [🎉️ Example](#%EF%B8%8F-example)
+- [🔍️ Reference](#%EF%B8%8F-reference)
+
+
+## 🌱 Prerequisites
+
+- This Executor works on Python 3.7 and 3.8.
```
Comment (tadejsv) [inline jinahub/rankers/LightGBMRanker/README.md:17]: And 3.9?

File: jinahub/rankers/LightGBMRanker/README.md [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+- [🌱 Prerequisites](#-prerequisites)
+- [🚀 Usages](#-usages)
+- [🎉️ Example](#%EF%B8%8F-example)
+- [🔍️ Reference](#%EF%B8%8F-reference)
+
+
+## 🌱 Prerequisites
+
+- This Executor works on Python 3.7 and 3.8. 
+- Make sure to install the [requirements](requirements.txt)
```
Comment (tadejsv) [inline jinahub/rankers/LightGBMRanker/README.md:18]: Should only be necessary when developing locally, not with `jinahub+docker` or `jinahub`, I believe

File: jinahub/rankers/LightGBMRanker/README.md [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+
+
+	
+
+## 🎉️ Example 
+
+```python
+from jina import Flow, DocumentArray
+
+f = Flow().add(uses='jinahub://LightGBMRanker')
```
Comment (tadejsv) [inline jinahub/rankers/LightGBMRanker/README.md:70]: Won't this example crash? You are not providing required arguments in the init function

File: jinahub/rankers/LightGBMRanker/lightgbm_ranker.py [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+
+import lightgbm
+import numpy as np
+from jina.logging.logger import JinaLogger
+from jina import Executor, DocumentArray, requests
+
+
+class LightGBMRanker(Executor):
+    """
+    Computes a relevance score for each match using a pretrained Ltr model trained with LightGBM (https://lightgbm.readthedocs.io/en/latest/index.html)
```
Comment (tadejsv) [inline jinahub/rankers/LightGBMRanker/lightgbm_ranker.py:15]:
```suggestion
    Computes a relevance score for each match using a pretrained learning-to-rank model trained with LightGBM (https://lightgbm.readthedocs.io/en/latest/index.html)
```

Comment (tadejsv) [reply jinahub/rankers/LightGBMRanker/lightgbm_ranker.py:15]: Also, add an empty line for readability here

File: jinahub/rankers/LightGBMRanker/tests/integration/flow.yml [hunk 7]
```diff
@@ -0,0 +1,15 @@
+jtype: Flow
+version: '1'
+with:
```
Comment (tadejsv) [inline jinahub/rankers/LightGBMRanker/tests/integration/flow.yml:3]: Can you put these arguments under the pod itself?

File: jinahub/rankers/LightGBMRanker/tests/integration/test_lightgbm.py [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+import os
+
+import pytest
+from jina import Flow
+
+cur_dir = os.path.dirname(os.path.abspath(__file__))
+
+
+@pytest.fixture
+def model_path(tmpdir):
```
Comment (tadejsv) [inline jinahub/rankers/LightGBMRanker/tests/integration/test_lightgbm.py:13]: Use `tmp_path`, `tmpdir` is based on a deprecated library + you can use `pathlib` methods with `tmp_path`

File: jinahub/rankers/LightGBMRanker/tests/integration/test_lightgbm.py [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+    It will be trained on fake data forcing that the first feature correlates positively with relevance, while the rest are
+    are random
+    :param tmpdir:
+    :return:
+    """
+    model_path = os.path.join(tmpdir, 'model.txt')
+    return model_path
+
+
+def test_train_offline(
```
Comment (tadejsv) [inline jinahub/rankers/LightGBMRanker/tests/integration/test_lightgbm.py:25]: This looks a bit complicated for an integration test

File: jinahub/rankers/LightGBMRanker/lightgbm_ranker.py [hunk 10]
```diff
@@ -0,0 +1,209 @@
...
+        """Load model from model path"""
+        self.booster = lightgbm.Booster(model_file=path)
+        model_num_features = self.booster.num_feature()
+        expected_num_features = len(self.query_features + self.match_features)
+        if self.categorical_query_features:
+            expected_num_features += len(self.categorical_query_features)
+        if self.categorical_match_features:
+            expected_num_features += len(self.categorical_match_features)
+        if model_num_features != expected_num_features:
+            raise ValueError(
```
Comment (tadejsv) [inline jinahub/rankers/LightGBMRanker/lightgbm_ranker.py:78]: Test this error

File: jinahub/rankers/LightGBMRanker/lightgbm_ranker.py [hunk 11]
```diff
@@ -0,0 +1,209 @@
...
+
+    def _get_features_dataset(self, docs: DocumentArray) -> 'lightgbm.Dataset':
+        q_features, m_features, group, labels = [], [], [], []
+        query_feature_names = self.query_features
+        if self.categorical_query_features:
+            query_feature_names += self.categorical_query_features
+        match_feature_names = self.match_features
+        if self.categorical_match_features:
+            match_feature_names += self.categorical_match_features
+        for doc in docs:
```
Comment (tadejsv) [inline jinahub/rankers/LightGBMRanker/lightgbm_ranker.py:95]: I would refactor this - use `docs.get_attributes(f'tags__{tag_name}')`, instead of looping over documents

File: jinahub/rankers/LightGBMRanker/lightgbm_ranker.py [hunk 12]
```diff
@@ -0,0 +1,209 @@
...
+    from model_path
+    :param match_features: name of the features to extract from match Documents and used to compute relevance scores by the model loaded
+    from model_path
+    :param relevance_label: If call :meth:`train` endpoint, the label will be used as groundtruth for model training. If on :meth:`rank` endpoint, the
+    label will be used to assign a score to :attr:`Document.scores` field.
+    :param model_path: path to the pretrained model previously trained using LightGBM.
+    :param params: Parameters used to train the LightGBM learning-to-rank model.
+    :param categorical_query_features: name of features contained in `query_features` corresponding to categorical features.
+    :param categorical_match_features: name of features contained in `match_features` corresponding to categorical features.
+    :param query_features_before: True if `query_features` must be placed before the `match` ones in the `dataset` used for prediction.
```
Comment (tadejsv) [inline jinahub/rankers/LightGBMRanker/lightgbm_ranker.py:26]: Is this really needed? I think we can just establish a convention, note it, and that's it. If we train the model in the executor it doesn't matter, if it is trained outside the users can easily set the proper order of features.

File: jinahub/rankers/LightGBMRanker/lightgbm_ranker.py [hunk 13]
```diff
@@ -0,0 +1,209 @@
...
+    def dump(self, parameters: Dict, **kwargs):
+        """Dump trained model to specified path
+
+        :param parameters: Parameters pass to this endpoint, expect :attr:`model_path` to be set.
+        """
+        model_path = parameters.get('model_path', None)
+        if model_path:
+            self._save_model(model_path)
+        else:
+            raise ValueError(
```
Comment (tadejsv) [inline jinahub/rankers/LightGBMRanker/lightgbm_ranker.py:193]: Test this error

File: jinahub/rankers/LightGBMRanker/lightgbm_ranker.py [hunk 14]
```diff
@@ -0,0 +1,209 @@
...
+        dataset = self._get_features_dataset(docs)
+        if self.booster:
+            predictions = self.booster.predict(dataset.get_data())
+            matches = docs.traverse_flat(traversal_paths=['m'])
+            for prediction, match in zip(predictions, matches):
+                match.scores[self.label] = prediction
+            for doc in docs:
+                doc.matches.sort(key=lambda x: x.scores[self.label].value, reverse=True)
+        else:
+            self.logger.warning('Model not trained, will return initial docs.')
```
Comment (tadejsv) [inline jinahub/rankers/LightGBMRanker/lightgbm_ranker.py:181]: Test this warning using `caplog` fixture (if it's not too much effort)

File: jinahub/rankers/LightGBMRanker/lightgbm_ranker.py [hunk 15]
```diff
@@ -0,0 +1,209 @@
...
+    def load(self, parameters: Dict, **kwargs):
+        """Load trained model from specified path
+
+        :param parameters: Parameters pass to this endpoint, expect :attr:`model_path` to be set.
+        """
+        model_path = parameters.get('model_path', self.model_path)
+        if model_path:
+            self._load_model(model_path)
+        else:
+            raise FileNotFoundError(
```
Comment (tadejsv) [inline jinahub/rankers/LightGBMRanker/lightgbm_ranker.py:207]: Test this error

**Review by tadejsv — CHANGES_REQUESTED** [review]

File: jinahub/rankers/LightGBMRanker/lightgbm_ranker.py (hunk 11 above)
Comment (tadejsv) [reply jinahub/rankers/LightGBMRanker/lightgbm_ranker.py:95]: Nevermind actually - I see you are doing it because of the matches

File: jinahub/rankers/LightGBMRanker/lightgbm_ranker.py (hunk 12 above)
Comment (JoanFM) [reply jinahub/rankers/LightGBMRanker/lightgbm_ranker.py:26]: Let's keep it like this and maybe change it later.

File: jinahub/rankers/LightGBMRanker/tests/integration/test_lightgbm.py (hunk 9 above)
Comment (JoanFM) [reply jinahub/rankers/LightGBMRanker/tests/integration/test_lightgbm.py:25]: Yes, integration test should be easy. Complex things should be tested in `unit` unless `reloading` mechanisms or so need to be tested which is not the case

File: jinahub/rankers/LightGBMRanker/README.md [hunk 16]
```diff
@@ -14,8 +14,8 @@
 
 ## 🌱 Prerequisites
 
-- This Executor works on Python 3.7 and 3.8. 
-- Make sure to install the [requirements](requirements.txt)
+- This Executor works on Python 3.7, 3.8 and 3.9. 
+- WHile developing locally, make sure to install the [requirements](requirements.txt)
```
Comment (JoanFM) [inline jinahub/rankers/LightGBMRanker/README.md:18]:
```suggestion
- While developing locally, make sure to install the [requirements](requirements.txt)
```

File: jinahub/rankers/LightGBMRanker/lightgbm_ranker.py [hunk 17]
```diff
@@ -0,0 +1,209 @@
...
+import lightgbm
+import numpy as np
+from jina.logging.logger import JinaLogger
+from jina import Executor, DocumentArray, requests
+
+
+class LightGBMRanker(Executor):
+    """
+    Computes a relevance score for each match using a pretrained learning-to-rank model trained with LightGBM (https://lightgbm.readthedocs.io/en/latest/index.html)
+    :param query_features: name of the features to extract from query Documents and used to compute relevance scores by the model loaded
```
Comment (tadejsv) [inline jinahub/rankers/LightGBMRanker/lightgbm_ranker.py:16]: add empty line for better readability
```suggestion

    :param query_features: name of the features to extract from query Documents and used to compute relevance scores by the model loaded
```

File: jinahub/rankers/LightGBMRanker/lightgbm_ranker.py [hunk 18]
```diff
@@ -0,0 +1,209 @@
...
+        self.params = params
+        self.model_path = model_path
+        self.query_features = query_features
+        self.match_features = match_features
+        self.categorical_query_features = categorical_query_features
+        self.categorical_match_features = categorical_match_features
+        self.query_features_before = query_features_before
+        self.label = relevance_label
+        self.logger = JinaLogger(self.__class__.__name__)
+        if self.model_path and os.path.exists(self.model_path):
```
Comment (tadejsv) [inline jinahub/rankers/LightGBMRanker/lightgbm_ranker.py:63]: How about we thrown an error, or at least a warning, if model path was provided but doesn't exist?

**Review by tadejsv — COMMENTED** [review]
Minor comments

File: jinahub/rankers/LightGBMRanker/lightgbm_ranker.py (hunk 14 above)
Comment (bwanglzu) [reply jinahub/rankers/LightGBMRanker/lightgbm_ranker.py:185]: i don't see it's necessary..

**Review by nan-wang — APPROVED** [review]
LGTM👍

**Review by tadejsv — APPROVED** [review]
Looks good to me

**Review by JoanFM — APPROVED** [review]
